File: frontend/plugins/location-sync/android/src/main/java/com/pathguard/app/plugin/LocationHttpClient.java
package com.pathguard.app.plugin;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.TimeUnit;

import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;

public class LocationHttpClient {
    private static final MediaType JSON = MediaType.parse("application/json; charset=utf-8");
    private final OkHttpClient client;

    public LocationHttpClient() {
        this.client = new OkHttpClient.Builder()
            .connectTimeout(10, TimeUnit.SECONDS)
            .writeTimeout(10, TimeUnit.SECONDS)
            .readTimeout(10, TimeUnit.SECONDS)
            .build();
    }

    public boolean sendBatch(List<LocationPoint> batch, int walkId, String deviceToken, String serverUrl, SimpleDateFormat isoFormatter) {
        JsonArray pointsArray = new JsonArray();
        for (LocationPoint p : batch) {
            JsonObject obj = new JsonObject();
            obj.addProperty("latitude", p.latitude);
            obj.addProperty("longitude", p.longitude);
            obj.addProperty("timestamp", isoFormatter.format(new Date(p.timestampMs)));
            obj.addProperty("client_id", p.clientId);
            obj.addProperty("walk_id", walkId);
            obj.addProperty("is_recovered", p.isRecovered);
            pointsArray.add(obj);
        }

        JsonObject body = new JsonObject();
        body.addProperty("walk_id", walkId);
        body.addProperty("batch_id", UUID.randomUUID().toString());
        body.add("points", pointsArray);

        String url = serverUrl.replaceAll("/$", "") + "/locations/batch";

        Request request = new Request.Builder()
            .url(url)
            .header("Content-Type", "application/json")
            .header("X-Patient-Token", deviceToken)
            .post(RequestBody.create(body.toString(), JSON))
            .build();

        try {
            okhttp3.Response response = client.newCall(request).execute();
            boolean success = response.isSuccessful() || response.code() == 409;
            response.close();
            return success;
        } catch (Exception e) {
            return false;
        }
    }
}

File: frontend/plugins/location-sync/ios/Plugin/LocationHttpClient.swift
import Foundation

class LocationHttpClient {
    private let session: URLSession
    private let timeout: TimeInterval = 15
    private let isoFormatter: ISO8601DateFormatter
    
    init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = timeout
        config.timeoutIntervalForResource = timeout
        self.session = URLSession(configuration: config)
        self.isoFormatter = ISO8601DateFormatter()
        self.isoFormatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    }
    
    func sendBatch(
        _ batch: [LocationPoint],
        walkId: Int,
        deviceToken: String,
        serverUrl: String,
        completion: @escaping (Bool) -> Void
    ) {
        var pointsArray: [[String: Any]] = []
        for p in batch {
            let date = Date(timeIntervalSince1970: TimeInterval(p.timestampMs) / 1000.0)
            let point: [String: Any] = [
                "latitude": p.latitude,
                "longitude": p.longitude,
                "timestamp": isoFormatter.string(from: date),
                "client_id": p.clientId,
                "walk_id": walkId,
                "is_recovered": p.isRecovered
            ]
            pointsArray.append(point)
        }
        
        let body: [String: Any] = [
            "walk_id": walkId,
            "batch_id": UUID().uuidString,
            "points": pointsArray
        ]
        
        guard let jsonData = try? JSONSerialization.data(withJSONObject: body) else {
            NSLog("[LocationHttpClient] Failed to serialize JSON")
            completion(false)
            return
        }
        
        let cleanedUrl = serverUrl.hasSuffix("/") ? String(serverUrl.dropLast()) : serverUrl
        let fullUrl = "\(cleanedUrl)/locations/batch"
        NSLog("[LocationHttpClient] Sending to URL: \(fullUrl)")
        guard let url = URL(string: fullUrl) else {
            NSLog("[LocationHttpClient] Invalid URL: \(fullUrl)")
            completion(false)
            return
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(deviceToken, forHTTPHeaderField: "X-Patient-Token")
        request.httpBody = jsonData
        
        let task = session.dataTask(with: request) { _, response, error in
            if let error = error {
                NSLog("[LocationHttpClient] Network error: \(error.localizedDescription)")
                completion(false)
                return
            }
            guard let httpResponse = response as? HTTPURLResponse else {
                NSLog("[LocationHttpClient] Invalid HTTP response")
                completion(false)
                return
            }
            let success = (200...299).contains(httpResponse.statusCode)
            NSLog("[LocationHttpClient] HTTP response: statusCode=\(httpResponse.statusCode), success=\(success)")
            completion(success)
        }
        task.resume()
    }
}
